src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -84,130 +103,182 @@ private static SequencePosition GetBufferCrossSegment(int startIndex, in Sequenc
-        private static SequencePosition SeekMultiSegment(object startSegment, int startIndex, object endSegment, int endPosition, long count, bool checkEndReachable)
+        private static SequencePosition SeekMultiSegment(ReadOnlySequenceSegment<T> currentSegment, int startIndex, object endObject, int endPosition, long count)
         {
-            var current = (IMemoryList<T>)startSegment;
-            var end = (IMemoryList<T>)endSegment;
+            Debug.Assert(currentSegment != null);
+            Debug.Assert(currentSegment.Next != null);
 
-            SequencePosition result = default;
-            int currentIndex = startIndex;
-
-            do
+            if (count == 0)
             {
-                bool isCurrentAtEnd = current == end;
-
-                int memoryLength = isCurrentAtEnd ? endPosition - currentIndex : current.Memory.Length - currentIndex;
+                // End of segment. Move to start of next.
+                currentSegment = currentSegment.Next;
+            }
+            else
+            {
+                // Not in this segment. Move to next
+                currentSegment = currentSegment.Next;
 
-                if (memoryLength > count || (memoryLength == count && isCurrentAtEnd))
+                while (currentSegment != null)
                 {
-                    result = new SequencePosition(current, currentIndex + (int)count);
-                    break;
+                    bool isCurrentAtEnd = currentSegment == endObject;
+                    int memoryLength = isCurrentAtEnd ? endPosition : currentSegment.Memory.Length;
+
+                    if (memoryLength > count || (memoryLength == count && isCurrentAtEnd))
+                    {
+                        // Fully contained in this segment; or at end of segment but there isn't a next one to move to
+                        break;
+                    }
+
+                    // Move to next
+                    count -= memoryLength;
+                    currentSegment = currentSegment.Next;
                 }
+            }
 
-                count -= memoryLength;
-                currentIndex = 0;
-                current = current.Next;
-            } while (current != null);
-
-            if (current == null)
+            // Hit the end of the segments but didn't reach the count
+            if (currentSegment == null)
                 ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
 
-            if (checkEndReachable)
+            return new SequencePosition(currentSegment, (int)count);
+        }
+
+        private static void CheckEndReachable(object startSegment, object endSegment)
+        {
+            Debug.Assert(startSegment != null);
+            Debug.Assert(startSegment is ReadOnlySequenceSegment<T>);
+            Debug.Assert(endSegment != null);
+
+            var current = Unsafe.As<ReadOnlySequenceSegment<T>>(startSegment);
+
+            while (current.Next != null)
```

Comment:
Doesn't this just need to make sure that endSegment exists within the chain but doesn't need to check that the last non-null segment is the end?

A -> B ->C -> D

If end is C doesn't this fail?

Reply:
Add a test for this?

Reply:
Yeah it fails

Reply:
Fixed added tests